esr_key_value_database_intrinsics.md

```diff
@@ -0,0 +1,264 @@
+
+### kv_erase
+
+`kv_erase` erases a key. If `contract` doesn't match the executing contract then it aborts the transaction. Else the
+function succeeds, even if the key didn't exist.
+
+If any iterators are positioned at the erased key, then `kv_erase` switches their status to `iterator_erased`.
+
+`kv_erase` clears the temporary data buffer.
+
+### kv_set
+
+`kv_set` sets a key-value pair. If `contract` doesn't match the executing contract then it aborts the transaction.
+If the contract exceeds available resources, it will fail at the appropriate time, similar to existing database
+primitives.
+
+If any iterators are positioned at the key, then `kv_set` switches their status to `iterator_ok`. This allows
+iterators to live through `kv_erase` followed by `kv_set` unharmed.
+
+`kv_set` clears the temporary data buffer.
+
+Consensus parameters limit the maximum key size and the maximum value size. `kv_set` aborts the transaction if
+the contract exceeds these limits.
+
+### kv_get
+
+`kv_get` checks the existence of a key. If the key doesn't exist, it returns `false`, clears the temporary data buffer,
+and sets `value_size` to 0. If the key does exist, it returns `true`, stores the value into the temporary data buffer,
+and sets `value_size` to the value size. Use `kv_get_data` to retrieve the value.
+
+## Iterator Intrinsics
+
+```c++
+uint32_t kv_it_create(uint64_t db, uint64_t contract, const char* prefix, uint32_t size);
+void     kv_it_destroy(uint32_t itr);
+it_stat  kv_it_status(uint32_t itr);
+int      kv_it_compare(uint32_t itr_a, uint32_t itr_b);
+int      kv_it_key_compare(uint32_t itr, const char* key, uint32_t size);
+it_stat  kv_it_move_to_oob(uint32_t itr);
+it_stat  kv_it_increment(uint32_t itr);
+it_stat  kv_it_decrement(uint32_t itr);
+it_stat  kv_it_lower_bound(uint32_t itr, const char* key, uint32_t size);
+it_stat  kv_it_key(uint32_t itr, uint32_t offset, char* dest, uint32_t size, uint32_t& copied);
+it_stat  kv_it_value(uint32_t itr, uint32_t offset, char* dest, uint32_t size, uint32_t& copied);
+```
+
+These intrinsics support iterating over ranges of key-value pairs. An iterator, during its lifetime:
+
+* Stays within a single database ID
+* Stays within a single contract's data
+* Stays within the subset of keys defined by the prefix passed to `kv_it_create`
+
+An iterator has a status (`it_stat`), which is one of the following:
+
+| Name             | Value | Description |
+|------------------|-------|-------------|
+|`iterator_ok`     | 0     | Iterator is positioned at a key-value pair |
+|`iterator_erased` | -1    | The key-value pair that the iterator used to be positioned at was erased |
+|`iterator_oob`    | -2    | Iterator is out-of-bounds |
+
+### kv_it_create
+
+```c++
+uint32_t kv_it_create(uint64_t db, uint64_t contract, const char* prefix, uint32_t size);
+```
+
+`kv_it_create` creates an iterator. The returned handle:
```

Comment:
It may be worth explicitly pointing out that it successfully returns an iterator if the `contract` account doesn't exist.

Reply:
Added